service/mapper/transitions.go

```diff
@@ -269,9 +269,116 @@ func (t *Transitions) ResumeIndexing(s *State) error {
+		return fmt.Errorf("unable to retrieve trie updates for block height %x", s.height)
+	}
+	s.updates = updates
+	log.Info().Int("updates", len(s.updates)).Msg("collected trie updates (registers) to be mapped")
+
+	// Now that we have collected TrieUpdates for the current block,
+	// we can create a path -> payload mapping that we will use to index to the local storage
+	s.status = StatusCollect
+	return nil
+}
+
+// CollectRegisters reads the payloads for the next block to be indexed from the state's forest, unless payload
+// indexing is disabled.
+func (t *Transitions) CollectRegisters(s *State) error {
+	if s.status != StatusCollect {
+		return fmt.Errorf("invalid status for collecting registers (%s)", s.status)
+	}
+
+	// If indexing payloads is disabled, we can bypass collection and indexing
+	// of payloads and just go straight to indexing chain data
+	if t.cfg.SkipRegisters {
+		s.status = StatusIndex
+		return nil
+	}
+	// collect paths/payload combinations
+	for _, update := range s.updates {
+		// guard for bootstrap case where t.updates.AllUpdates() returns nil as the queue is empty
+		if update != nil {
+			for i, path := range update.Paths {
+				s.registers[path] = update.Payloads[i]
+			}
+		}
+	}
+
+	// At this point, we have collected all the payloads, so we go to the next
+	// step, where we will index them.
+	s.status = StatusMap
+	return nil
+}
+
+// MapRegisters maps the collected registers to the current block.
+func (t *Transitions) MapRegisters(s *State) error {
+	if s.status != StatusMap {
+		return fmt.Errorf("invalid status for indexing registers (%s)", s.status)
+	}
+
+	log := t.log.With().Uint64("height", s.height).Logger()
+	// If there are no registers left to be indexed, we can go to the next step,
+	// which is indexing the chain data
+	if len(s.registers) == 0 {
+		s.status = StatusIndex
+		return nil
+	}
+
+	// We will now collect and index 1000 registers at a time. This gives the
+	// FSM the chance to exit the loop between every 1000 payloads we index. It
+	// doesn't really matter for badger if they are in random order, so this
+	// way of iterating should be fine.
+	n := 1000
+	payloads := make([]*ledger.Payload, 0, n)
+	for path, payload := range s.registers {
+		payloads = append(payloads, payload)
+		delete(s.registers, path)
+		if len(payloads) >= n {
+			break
+		}
+	}
+
+	// Then we store the (maximum) 1000 paths and payloads.
+	err := t.write.Payloads(s.height, payloads)
+	if err != nil {
+		return fmt.Errorf("could not index registers: %w", err)
+	}
+
+	// skip the map status again, and its log
+	if len(s.registers) == 0 {
+		// write the height as latest
+		err = t.write.LatestRegisterHeight(s.height)
+		if err != nil {
+			return fmt.Errorf("could not index latest register height: %w", err)
+		}
+		s.status = StatusIndex
+		return nil
+	}
+
+	log.Debug().
+		Int("batch", len(payloads)).
+		Int("remaining", len(s.registers)).
+		Msg("indexed register batch for finalized block")
+
 	return nil
```

Comment:
why `LatestRegisterHeight` and `s.status = StatusIndex` are not needed here?

Reply:
updated in line `369` but at this part of the code, we may still have some batches of registers remaining for the current block, so we don't update it yet.